lex/lexer.go

```diff
@@ -55,38 +55,69 @@ const (
 	hexDigits = "0123456789ABCDEF"
 )
 
-// StateFn represents the state of the lexer as a function that returns the
-// next state.
-type StateFn func(*Lexer) StateFn
+type (
+	// StateFn represents the state of the lexer as a function that returns the
+	// next state.
+	StateFn func(*Lexer) StateFn
 
-// NamedStateFn is a StateFn which has a name for tracing debugging.
-type NamedStateFn struct {
-	Name    string
-	StateFn StateFn
+	// NamedStateFn is a StateFn which has a name for tracing debugging.
+	NamedStateFn struct {
+		Name    string
+		StateFn StateFn
+	}
+
+	LexerOption func(*Lexer)
+)
+
+func WithInput(input string) LexerOption {
+	return func(l *Lexer) {
+		l.input = input
+	}
+}
+
+func WithDialect(dialect *Dialect) LexerOption {
+	return func(l *Lexer) {
+		l.dialect = dialect
+		if len(dialect.IdentityQuoting) > 0 {
+			l.identityRunes = dialect.IdentityQuoting
+		} else {
+			l.identityRunes = IdentityQuoting
+		}
+	}
 }
 
-// NewLexerState Creates a new lexer for the input string and initial state.
-func NewLexerState(input string, dialect *Dialect, initialState StateFn) *Lexer {
-	// Three tokens of buffering is sufficient for all state functions.
+func WithInitialState(initialState StateFn) LexerOption {
+	return func(l *Lexer) {
+		l.state = initialState
+	}
+}
+
+func WithOverrideDepthLimit(override bool) LexerOption {
```

Comment:
seems like it might be nice if this option took an integer size, so you could reset the limit to a custom value, and maybe anything < 0 disables it entirely>

Reply:
^yeah I agree

Reply:
thanks, yeah good to have option to provide limit.